### PR TITLE
Always await step.value

### LINKED ISSUE #14470: [Bug]: for await loop transpiled by @babel/plugin-proposal-async-generator-functions does not await step.value
### 💻

- [X] Would you like to work on a fix?

### How are you using Babel?

Programmatic API (`babel.transform`, `babel.parse`)

### Input code

Repl: https://babeljs.io/repl#?browsers=defaults%2C%20not%20ie%2011%2C%20not%20ie_mob%2011&build=&builtIns=false&corejs=3.21&spec=false&loose=false&code_lz=IYZwngdgxgBAZgV2gFwJYHsICoaklAWWAAcAKEdBAJygFMAaeCAShgG8AoGedK3Ad2CpkMUlEwgRAN2AAbBLRjo4MCtTqtO3bmFS1ZAEyakZ82swDcXGAF8OdjlFmgQMAmACC4aAHFaEWipgZF52awCAD2RSTWtuKlpkagh2GANMWgAueDkQBhhTBWyABSp0AFtUPIA6BIpZKVpSACZWO252mGsAbQBlMHKAI3RZarxoAElkQODeAF0YsO0YBKSqFOQACyqrDvsORwkRNgK5BUZ0gNsYAF4BIRFxwhJSAP43T28oPwCgkKoYowIrcAHwwYE4VrVSLRSwHAD08NOZluNzuADlgOigA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=true&fileSize=false&timeTravel=false&sourceType=module&lineWrap=true&presets=env%2Creact%2Cstage-2&prettier=false&targets=&version=7.17.9&externalPlugins=%40babel%2Fplugin-proposal-async-generator-functions%407.16.8&assumptions=%7B%7D

### Configuration file name

babel.config.json

### Configuration
```
{
  "plugins": ["@babel/plugin-proposal-async-generator-functions"]
}
```

### Current and expected behavior

Current behavior:
A `for await .. of` loop inside an async generator will await a `step` (the result of `iterator.next()`) but `step.value` is not awaited. The plugin generates something like this:
```
for (var _step; (_step = yield _awaitAsyncGenerator(_iterator.next())).done;) {
  var value = step.value;
}
```

Expected behavior:
A `for await .. of` inside an async generator will await `step` and `step.value`. The plugin should generate something like this:
```
for (var _step; (_step = yield _awaitAsyncGenerator(_iterator.next())).done;) {
  var value = yield _awaitAsyncGenerator(step.value);
}
```

### Environment

All

### Possible solution

Fix the plugin to generate spec-compliant code

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by conartist6**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            |  Fixes #14470
| Patch: Bug Fix?          | 👍
| Major: Breaking Change?  | No
| Minor: New Feature?      |
| Tests Added + Pass?      | Yes
| Documentation PR Link    | none
| Any Dependency Changes?  | No
| License                  | MIT

Ensure that during async iteration `step.value` is always awaited. This includes `for await .. of` loops and `yield*` expressions. Currently this works for sync iterators being promoted to async iterators, but the spec dictates that async iterators may also return steps where `step.value instanceof Promise`, and those must be awaited. The current implementation does not await them, and in fact tests that it does not!

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/14471"></a>

### CODE REVIEW

**Comment by conartist6** [comment]
Not sure why node 8 would be failing yet... Will investigate.

**Comment by conartist6** [comment]
Ah, node <= 8 doesn't have `Symbol.asyncIterator`. I changed something I didn't understand, and the change was incorrect. I put it back.

**Comment by conartist6** [comment]
It's remarkably tricky to read some of this code that's written in styles you'd never write in otherwise, especially if you spend a lot of time using babel. I believe everything is now correct. I'm still looking for the part of the spec that defines these behaviors, but I already know what it says, and I've double checked that these changes are consistent with browser behavior.

**Comment by conartist6** [comment]
OK, and here's my investigation of the spec. The relevant parts are: [the AsyncIterator interface](https://tc39.es/ecma262/multipage/control-abstraction-objects.html#sec-asynciterator-interface), and [ForIn/OfBodyEvalutation](https://tc39.es/ecma262/multipage/ecmascript-language-statements-and-declarations.html#sec-runtime-semantics-forin-div-ofbodyevaluation-lhs-stmt-iterator-lhskind-labelset).

Aaaand now that I look the spec says I'm wrong and every browser I can test works the way babel does. Um, I guess I learned something today!

**Comment by conartist6** [comment]
Waaaait a second. JSfiddle is probably using babel.

**Comment by conartist6** [comment]
Yep, it was. Testing reveals that v8 and spidermonkey both do the more permissive thing that this PR implements.

**Comment by conartist6** [comment]
My new test is:
```html
<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8" />
    <script>
      class MyAsyncGenerator {
        next() {
          return { done: false, value: Promise.resolve(2) }
        }
        
        [Symbol.iterator]() {
          return this;
        }
      }

      async function test(source) {
        let i = 0;
        for await (const value of source) {
          if (i++ > 2) break; 
          console.log(value.toString());
        }
      }

      test(new MyAsyncGenerator())
    </script>
  </head>
</html>
```

**Comment by conartist6** [comment]
Looks to me like there was a flaky failure in the `babel-cli` test. Trying CI validation again.

**Comment by nicolo-ribaudo** [comment]
Yup, `babel-cli` has a flaky test. Sorry for that!

**Comment by nicolo-ribaudo** [comment]
I remember already fixing something similar to this in the past, I'm trying to find it to see if there was any relevant discussion.

File: packages/babel-helpers/src/helpers.ts [hunk 1]
```diff
@@ -126,11 +126,18 @@ helpers.asyncGeneratorDelegate = helper("7.0.0-beta.0")`
...
     var iter = {}, waiting = false;
 
-    function pump(key, value) {
-      waiting = true;
-      value = new Promise(function (resolve) { resolve(inner[key](value)); });
-      return { done: false, value: awaitWrap(value) };
-    };
+    function pump(key, arg) {
+      var step = inner[key](arg);
+      if (step instanceof Promise) {
```
Comment (nicolo-ribaudo) [inline packages/babel-helpers/src/helpers.ts:131]: I'm really surprised that we need to check if it's a promise and not if it's a thenable.

**Comment by nicolo-ribaudo** [comment]
Ok, it was https://github.com/babel/babel/pull/13491

File: packages/babel-helpers/src/helpers.ts (hunk 1 above)
Comment (conartist6) [reply packages/babel-helpers/src/helpers.ts:131]: Oh I have no idea. You think thenable is right? It probably is.

**Comment by conartist6** [comment]
Huh. So you changed the loop from
```js
for (
  var _iterator = _asyncIterator(y), _step, _value;
  _step = yield _wrapAwait(_iterator.next()),
    _iteratorNormalCompletion = _step.done,
    _value = yield _wrapAwait(_step.value),
    !_iteratorNormalCompletion;
  _iteratorNormalCompletion = true
) {
  let x = _value;
  f(x);
}
```
to
```js
for (
  var _iterator = _asyncIterator(y), _step;
  _iteratorAbruptCompletion = !(_step = yield _wrapAwait(_iterator.next())).done;
  _iteratorAbruptCompletion = false
) {
  let x = _step.value;
  f(x);
}
```

**Comment by conartist6** [comment]
I agree that this is compliant with the spec, but I don't understand the relationship between the original bug and its fix. In my testing browsers (currently) do what this PR does, so any deviation from that behavior will lead to possible unexpected issues when going from a native environment to a transpiled one.

**Comment by nicolo-ribaudo** [comment]
What should https://github.com/babel/babel/pull/14471#issuecomment-1100910503 print? I tested it both in Firefox and compiled with Babel (with the current Babel version, I didn't test this PR), and it logs `2` in both cases. I also tested https://babeljs.io/repl#?browsers=defaults%2C%20not%20ie%2011%2C%20not%20ie_mob%2011&build=&builtIns=false&corejs=3.21&spec=false&loose=false&code_lz=IYZwngdgxgBAZgV2gFwJYHsICoaklAWWAAcAKEdBAJygFMAaeCAShgG8AoGedK3Ad2CpkMUlEwgRAN2AAbBLRjo4MCtTqtO3bmFS1ZAEyakZ82swDcXGAF8OdjlFmgQMAmACC4aAHFaEWipgZF52awCAD2RSTWtuKlpkagh2GANMWgAueDkQBhhTBWyABSp0AFtUPIA6BIpZKVpSACZWO252mGsAbQBlMHKAI3RZarxoAElkQODeAF0YsO0YBKSqFOQACyqrDvsORwkRNgK5BUZ0gNsYAF4BIRFxwhJSAP43T28oPwCgkKoYowIrcAHwwYE4VrVSLRSwHAD08NOZluNzuADlgOigA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=true&fileSize=false&timeTravel=false&sourceType=module&lineWrap=true&presets=env%2Creact%2Cstage-2&prettier=false&targets=&version=7.17.9&externalPlugins=%40babel%2Fplugin-proposal-async-generator-functions%407.16.8&assumptions=%7B%7D, and it returns `{ value: NaN, done: false }` both with Babel and without.

What is an example that is currently compiled wrongly by Babel?

**Comment by conartist6** [comment]
Logging `2` is correct, as opposed to `Promise {<fulfilled>: 2}` which is what I have observed babel to do.

**Comment by nicolo-ribaudo** [comment]
Uhm, I see `2` :sweat_smile: 

https://babeljs.io/repl#?browsers=defaults%2C%20not%20ie%2011%2C%20not%20ie_mob%2011&build=&builtIns=false&corejs=3.21&spec=false&loose=false&code_lz=MYGwhgzhAECyCeBBC8B2wDiBTVWBOYALgPZ7QDeAUNNLgB6EAUAlBdTdHloQK56oVoAE2K4AXNABmYEBCwAaaADcZPLBIAKeYgFsAlnIB0XCMRBKsjAEysAvu3s12AbQDK8HQCMzhvYXxEpAC6LGwcnNx8AoQAFgYA3A6U9pSQaMBSPOiEeqLQ_hBMpnzAWKxUNCDc0HrQALzQAAyJNJKk0GAA7mB-0IzAooXKqljQxJLQxXil5ew0ehOMegDUy9AAfA0AjKyeXGAA1vHQcx1VeEwqIGqGJK6EeHqoAOYszC3Q9ikFTLidcEgUOhsLgCCQ8G8gA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=true&fileSize=false&timeTravel=false&sourceType=module&lineWrap=true&presets=env%2Creact%2Cstage-2&prettier=false&targets=&version=7.17.9&externalPlugins=%40babel%2Fplugin-proposal-async-generator-functions%407.16.8&assumptions=%7B%7D

(I changed `console.log` to `alert`, but you probably need to deselect and re-select "Evaluate" on the left sidebar)

Maybe you are using an old Babel version? I also found https://github.com/babel/babel/pull/13824, so it looks like versions >=17.14.9 <17.16.0 are broken.

**Comment by conartist6** [comment]
Oh oops, that's slightly wrong. As I think I mentioned somewhere it works fine for iterators, but not asyncIterators. To break it change
```js
  [Symbol.iterator]() {
    return this;
  }
  //  to
    [Symbol.asyncIterator]() {
    return this;
  }
```

**Comment by conartist6** [comment]
The spec says "Additionally, the IteratorResult object that serves as a fulfillment value should have a "value" property whose value is not a promise (or "thenable"). However, this requirement is also not enforced."

Given that the spec admits that it's entirely possible that value will be a promise, and that a promise would have special meaning as a value, I think that is enough to make it a condition that needs to be handled. Browser implementers seem to have thought so, and I'd say it follows the principle of least surprise for coders too. The syntax reads "for await value of iterable". To me that suggests that by the time I get a value it has been awaited, and I would never need to await it again for defensive purposes.

**Comment by nicolo-ribaudo** [comment]
If I use `Symbol.asyncIterator`, it alerts `[object Promise]` both with and without Babel.

> The spec says "Additionally, the IteratorResult object that serves as a fulfillment value should have a "value" property whose value is not a promise (or "thenable"). However, this requirement is also not enforced."

Note that "this requirement is also not enforced" does not meant that browsers can do what they want (the browser behavior is well-defined in the algorithm steps), but that JS _users_ can write code such that `value` is a promise.

**Comment by conartist6** [comment]
I've not observed it to do that without babel. How are you testing? Here's a test page: https://burningpotato.com/for-await-test.html

**Comment by conartist6** [comment]
D'oh, and of course the test I uploaded was not making an `asyncIterator`.

**Comment by conartist6** [comment]
I guess the moral of the story is that I read babel's behavior from a long time ago, thought that was what the spec said, and then tested wrong so that I believed that browsers were doing what babel used to do, which they aren't.

**Comment by nicolo-ribaudo** [comment]
And the other part of the moral of the story is that this is a particularly complex and confusing topic! :stuck_out_tongue:

**Comment by conartist6** [comment]
It is indeed! If you're interested, this was all about perf testing and perhaps a proposal for some new language mechanics to radically improve the performance of async iteration of character streams, which are usually async iterables of string chunks. I'm able to show that skipping awaits on synchronously available steps by modifying the code for this plugin can result in a 4x speedup for a small parser. The test cases and results are [here](https://github.com/conartist6/async-perf)

**Pull request closed without merging**